File: 1DDynamicProgramming/cppRealizations/palindromicSubstrings.cpp
#include <string>
#include <iostream>

using namespace std;


class Solution 
{
public:
    int countSubstrings(string s) 
    {
        int nSubstrings = 0;

        for(int i = 0; i < s.size(); i++)
        {
            // odd length
            int left = i, right = i;
            while(left >= 0 && right < s.size() && s[left] == s[right])
            {
                nSubstrings++;
                left--;
                right++;
            }

            // even length
            left = i, right = i + 1;
            while(left >= 0 && right < s.size() && s[left] == s[right])
            {
                nSubstrings++;
                left--;
                right++;
            }
        }
        return nSubstrings;
    }
};


int main()
{
    string s = "abc";
    cout << "Input: " << s << endl;
    cout << "Output: " << Solution().countSubstrings(s) << endl;
    return 0;
}

File: 1DDynamicProgramming/palindromicSubstrings.py
class Solution:
    def countSubstrings(self, s: str) -> int:
        nSubstrings = 0
        for i in range(len(s)):
            # odd len
            l, r = i, i
            while l >= 0 and r < len(s) and s[l] == s[r]:
                nSubstrings += 1
                l -= 1
                r += 1

            # even len
            l, r = i, i + 1
            while l >= 0 and r < len(s) and s[l] == s[r]:
                nSubstrings += 1
                l -= 1
                r += 1
        
        return nSubstrings


if __name__ == '__main__':
    input = "abc"
    print("Input: {}".format(input))
    print("Output: {}".format(Solution().countSubstrings(input)))